### PR TITLE
Feature/edit company view

### PR DESCRIPTION
**Pull request by NataliaPonomarevaMM**

### CODE REVIEW

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 1]
```diff
@@ -1,133 +1,104 @@
...
         return <View>
             <ScrollView 
                 horizontal 
                 pagingEnabled
-                contentOffset={{x: Dimensions.get('window').width * (this.props.departmentsLists !== undefined ? this.props.departmentsLists.currentPage : departments.length), y: 0}}
                 showsHorizontalScrollIndicator={false}
-                onMomentumScrollEnd={this.onMomentumScrollEnd}
-                onScrollBeginDrag={this.onScrollBeginDrag}
-                ref={ref => this.scrollView = ref as ScrollViewComponent}
+                onMomentumScrollEnd={this.onMomentumScrollEnd.bind(this)}
```
Comment (santee) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: revert this change and have callbacks as `private onMomentumScrollEnd = (event: NativeSyntheticEvent<NativeScrollEvent>) => {` ? Lambdas already do this bind for you

**Comment by serjKim** [comment]
I see a lot of looped api calls api/employees?departmentId=56 when click on the company tab.

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 2]
```diff
@@ -1,135 +1,118 @@
...
-            const nextEmployees = nextProps.employees.employeesById.filter(nextProps.employeesPredicate);	
+    private scrollView: ScrollViewComponent;
 
-            return !employees.equals(nextEmployees);
-        }	
+    public componentDidMount() {
+        this.setScroll();
+    }
+    public componentDidUpdate(prevProps: DepartmentsHScrollableListProps) {
+        if (!Set(prevProps.departments).equals(Set(this.props.departments)) ||
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: I guess Set(prevProps.departments) and Set(this.props.departments) are always not equal. It will work as you're expecting if departments are immutablejs-like objects

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 3]
```diff
@@ -1,135 +1,118 @@
...
         return <View>
             <ScrollView 
                 horizontal 
                 pagingEnabled
-                contentOffset={{x: Dimensions.get('window').width * (this.props.departmentsLists !== undefined ? this.props.departmentsLists.currentPage : departments.length), y: 0}}
                 showsHorizontalScrollIndicator={false}
                 onMomentumScrollEnd={this.onMomentumScrollEnd}
                 onScrollBeginDrag={this.onScrollBeginDrag}
-                ref={ref => this.scrollView = ref as ScrollViewComponent}
+                ref={component => this.scrollView = component as ScrollViewComponent}
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]:
```typescript
import { ..., ScrollView, ScrollViewProps } from 'react-native';
import { ..., Ref } from 'react';
...
private scrollView: ScrollView;
private ref: Ref<ScrollView & Component<ScrollViewProps>> = (instance) => {
     this.scrollView = instance;
}
...
<ScrollView ref={this.ref} ... />
```

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 4]
```diff
@@ -1,135 +1,118 @@
...
-                }
             </ScrollView>
         </View>;
     }
 
+    private setScroll() {
+        const x: number = this.props.departmentsLists ? this.props.departmentsLists.currentPage : 0;
+        const curOffsetX = Dimensions.get('window').width * x;
+
+        setTimeout(() => {
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: What's the point using of setTimeout?

File: mobile/src/people/departments/employee-card-with-avatar-list.tsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+interface EmployeeCardWithAvatarProps {
+    employees: Employee[];
+    treeLevel: number;
+    onItemClicked: (e: Employee) => void;
+}
+
+export class EmployeeCardWithAvatarList extends Component<EmployeeCardWithAvatarProps> {
+    public render() {
+        const calcultatedHeight = Dimensions.get('screen').height - 
+            (StyleSheet.flatten(styles.layout).height as number) * this.props.treeLevel - 119;
```
Comment (serjKim) [inline mobile/src/people/departments/employee-card-with-avatar-list.tsx]: Magic number

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 6]
```diff
@@ -1,135 +1,127 @@
...
-    employeesPredicate: (employee: Employee) => boolean;
-}
-
-interface ScrollViewComponent extends Component {
-    scrollTo(y?: number | { x?: number; y?: number; animated?: boolean }, x?: number, animated?: boolean): void;
-    scrollToEnd(): void;
+    key: string;
 }
 
 export interface DepartmentsListStateDescriptor {
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: Move this interface declaration to the people.reducer

File: mobile/src/reducers/organization/department.model.ts [hunk 7]
```diff
@@ -26,4 +26,8 @@ export class Department {
     @dataMember()
     @required({ nullable: true })
     public isHeadDepartment: boolean;
+
+    public equals(obj: Department): boolean {
```
Comment (serjKim) [inline mobile/src/reducers/organization/department.model.ts]: equals means when all instance's properties are equal to obj's properties, not only id

File: mobile/src/reducers/organization/employee.model.ts [hunk 8]
```diff
@@ -66,4 +66,8 @@ export class Employee {
     @dataMember()
     @required({nullable: true})
     public roomNumber: string;
+
+    public equals(obj: Employee): boolean {
```
Comment (serjKim) [inline mobile/src/reducers/organization/employee.model.ts]: Same

File: mobile/src/reducers/organization/department.model.ts [hunk 9]
```diff
@@ -26,4 +26,13 @@ export class Department {
...
```
Comment (serjKim) [inline mobile/src/reducers/organization/department.model.ts]: In beginning must be
```
if (!obj) return false;
if (obj === this) return true;
```

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 10]
```diff
@@ -1,135 +1,122 @@
...
-                        leftNeighbor={(index > 0 && subDepartments.length > 1) ? this.props.employees.employeesById.get(subDepartments[index - 1].chiefId) : null } 
-                        rightNeighbor={(index < subDepartments.length - 1 && subDepartments.length > 1) ? this.props.employees.employeesById.get(subDepartments[index + 1].chiefId) : null} 
+                        departmentAbbreviation={subDepartment.abbreviation}
+                        leftNeighbor={(index > 0) ? this.props.employees.employeesById.get(subDepartments[index - 1].chiefId) : null } 
+                        rightNeighbor={(index < subDepartments.length - 1) ? this.props.employees.employeesById.get(subDepartments[index + 1].chiefId) : null} 
                         ref={(employeeCard) => { 
-                            if ( employeeCard != null ) {
-                                this.employeeCards.push(employeeCard);
+                            if (employeeCard != null) {
+                                cur.employeeCards.push(employeeCard);
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: Move to a lambda method

File: mobile/src/people/departments/departments-horizontal-scrollable-list.tsx [hunk 11]
```diff
@@ -1,135 +1,122 @@
...
-        const { headDepartmentId, headDepartmentChiefId, departments } = this.props;
-        const subDepartments = headDepartmentId != null && departments.length > 0 ? departments : null;
-        let subordinates;
-        
-        if (this.props.employees.employeeIdsByDepartment.has(headDepartmentId) && this.props.employees.employeeIdsByDepartment.get(headDepartmentId).size > 0) {
-            subordinates = this.props.employees.employeesById.filter(this.props.employeesPredicate).toArray();
-        } else {
-            subordinates = null;
-        }
+        const cur = this;
```
Comment (serjKim) [inline mobile/src/people/departments/departments-horizontal-scrollable-list.tsx]: Not needed, just use lambda

File: mobile/src/reducers/search/search.epics.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+            if (curDep && !deps.find(e => e === curDep)) {
+                deps.push(curDep);
+            }
+        }
+    });
+    return deps;
+}
+
+export const updateDepartmentsBranchEpic$ = (action$: ActionsObservable<SetFilter>, appState: MiddlewareAPI<AppState>) =>
+    action$.ofType('SEARCH-BY-TEXT-FILTER')
```
Comment (serjKim) [inline mobile/src/reducers/search/search.epics.ts]: You should filter by `searchType = People` as well

File: mobile/src/reducers/search/search.epics.ts [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    .map(action => {
```
Comment (serjKim) [inline mobile/src/reducers/search/search.epics.ts]: What's the point having this logic in epic? can this logic be moved to reducer?

**Comment by serjKim** [comment]
:fire::fire::fire: Updated :fire::fire::fire:
